### PR TITLE
[Draft] Add PushChannelRequest Unittests

### PR DESCRIPTION
**Pull request by sharath2727**
Add channel request unittests

### CODE REVIEW

File: test/PushNotificationsTests/APITests.cpp [hunk 1]
```diff
@@ -29,9 +29,9 @@ namespace Test::PushNotifications
     private:
         wil::unique_event m_failed;
 
-        const std::wstring c_testPackageFile = g_deploymentDir + L"PushNotificationsTestPackage.msixbundle";
-        const std::wstring c_testPackageCertFile = g_deploymentDir + L"PushNotificationsTestPackage.cer";
-        const std::wstring c_testPackageFullName = L"PushNotificationsWin32App_1.0.0.0_x64__9eg3vg7cmy3rj";
+        const std::wstring c_testPackageFile = g_deploymentDir + L"MSIXPackager_1.0.8.0_x64.msixbundle";
```
Comment (danielayala94) [inline test/PushNotificationsTests/APITests.cpp:32]: Do we have a dependency on the app version? If yes, should we change the value of the const strings every time we increase the app version number?

File: test/PushNotificationsTests/APITests.cpp [hunk 2]
```diff
@@ -91,13 +91,65 @@ namespace Test::PushNotifications
...
             wil::unique_event event = CreateTestEvent(c_testProtocolScheme_Packaged);
 
             // This is associated protocol for the MSIX installed app for launch.
             // Use the ://path to define the component you want to test.
-            Uri launchUri{ c_testProtocolScheme_Packaged + L"://this_is_a_test" };
+            Uri launchUri{ c_testProtocolScheme_Packaged + L"://ChannelRequestUsingNullRemoteId" };
+            auto launchResult = Launcher::LaunchUriAsync(launchUri).get();
+            VERIFY_IS_TRUE(launchResult);
+
+            WaitForEvent(event, m_failed);
```
Comment (danielayala94) [inline test/PushNotificationsTests/APITests.cpp:104]: I'm seeing that we are launching an app where API testing will happen, or at least this is my expectation. Please correct me if this is wrong.

If I'm right, then where is the testing code?

File: test/TestApps/MSIXPackager/Package.appxmanifest [hunk 3]
```diff
@@ -8,8 +8,8 @@
 
   <Identity
     Name="PushNotificationsWin32App"
-    Publisher="CN=WNP_Reunion"
-    Version="1.0.0.0" />
+    Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US"
```
Comment (danielayala94) [inline test/TestApps/MSIXPackager/Package.appxmanifest:11]: Why a change here was needed? It affects package signing, right?

Comment (sharath2727) [reply test/TestApps/MSIXPackager/Package.appxmanifest:11]: Hi, as paul mentioned in one of the team meetings we are generating the certificate using tools/DevCheck.cmd file which I think is the right approach going forward for automation testing too. @pmpurifoy has more information on this. Thanks.

File: test/PushNotificationsTests/APITests.cpp (hunk 2 above)
Comment (sharath2727) [reply test/PushNotificationsTests/APITests.cpp:104]: test/TestApps/PushNotificationsTestApp/main.cpp -  can you see take a look this file? Thanks.

File: test/PushNotificationsTests/APITests.cpp (hunk 1 above)
Comment (sharath2727) [reply test/PushNotificationsTests/APITests.cpp:32]: Yeah. We need to. That is the way it works right now. But I think we can improvise this to look for a single version under the deploymentdir. That way we don't have to change the version number. Thanks.

File: test/PushNotificationsTests/APITests.cpp (hunk 2 above)
Comment (danielayala94) [reply test/PushNotificationsTests/APITests.cpp:104]: Ah GitHub didn't load it while reviewing. Thanks!

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 4]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    catch (...)
+    {
+        auto channelRequestException = hresult_error(to_hresult());
+        hr = channelRequestException.code();
+    }
+
+    if (hr == E_INVALIDARG)
+        return true;
+    else
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:54]: `return hr == E_INVALIDARG;`

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 5]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
-        // Post the JSON, and wait for a response.
-        httpResponseMessage = httpClient.PostAsync(
-            requestUri,
-            jsonContent).get();
-
-        // Make sure the post succeeded, and write out the response.
-        httpResponseMessage.EnsureSuccessStatusCode();
-        httpResponseBody = httpResponseMessage.Content().ReadAsStringAsync().get();
-        std::wcout << httpResponseBody.c_str() << std::endl;
+    // winrt::guid remoteId { L"a2e4a323-b518-4799-9e80-0b37aeb0d225" };
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:59]: Move closer to `channelOperation`.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 6]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(winrt::guid(0xa2e4a323, 0xb518, 0x4799, { 0x9e, 0x80, 0x0b, 0x37, 0xae, 0xb0, 0xd2, 0x25 }));
+
+    channelOperation.Progress(
+        [](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& /* sender */,
+            PushNotificationCreateChannelStatus const& args)
+        {
+            if (args.status == PushNotificationChannelStatus::InProgress)
+            {
+                // This is basically a noop since it isn't really an error state
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:71]: Nit: comment not needed.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 7]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+                std::cout << "The channel request is in progress " << args.extendedError << std::endl;
+            }
+            else if (args.status == PushNotificationChannelStatus::InProgressRetry)
+            {
+                std::cout << "The channel request is in back-off retry mode because of a retryable error code: " << args.extendedError << " Expect delays in acquiring it" << std::endl;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:76]: > Expect delays in acquiring it

What does this mean?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 8]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+
+    channelOperation.Progress(
+        [](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& /* sender */,
+            PushNotificationCreateChannelStatus const& args)
+        {
+            if (args.status == PushNotificationChannelStatus::InProgress)
+            {
+                // This is basically a noop since it isn't really an error state
+                std::cout << "The channel request is in progress " << args.extendedError << std::endl;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:72]: Nit:

`progress.`

or

`progress. Error code:`

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 9]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+        [&channelEvent](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
+            auto result = sender.GetResults();
+            if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+            {
+                auto channelUri = result.Channel().Uri();
+                auto channelExpiry = result.Channel().ExpirationTime();
+
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:91]: Nit: remove extra line.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 10]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+
+                std::cout << "channelUri: " << winrt::to_string(channelUri.ToString()) << std::endl;
+
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:94]: Nit: remove extra line.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 11]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    // Setup the completed event handler
+    channelOperation.Completed(
+        [&channelEvent](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
+            auto result = sender.GetResults();
+            if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+            {
+                auto channelUri = result.Channel().Uri();
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:89]: Should we verify if Channel URI is not empty?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 12]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+                auto channelExpiry = result.Channel().ExpirationTime();
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:90]: Should we print this or do something more than just calling the API? It is not used.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 13]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+
+                std::cout << "channelUri: " << winrt::to_string(channelUri.ToString()) << std::endl;
+
+            }
+            else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
+            {
+                std::cout << "Failed to complete the async complete handler, gracefully cancel the channelOperation" << std::endl;
+            }
+
+
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:101]: Don't leave two empty lines randomly. Please be consistent.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 14]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+                auto channelUri = result.Channel().Uri();
+                auto channelExpiry = result.Channel().ExpirationTime();
+
+
+                std::cout << "channelUri: " << winrt::to_string(channelUri.ToString()) << std::endl;
+
+            }
+            else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
+            {
+                std::cout << "Failed to complete the async complete handler, gracefully cancel the channelOperation" << std::endl;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:98]: Don't we want to print an error code?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 15]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
     {
-        std::wcout << ex.message().c_str() << std::endl;
+        std::cout << "Channel complete handler has been completed, safely closing the async operation" << std::endl;
+        channelOperation.Close(); // Do not call getresults after this
+    }
+
+    if (hr == S_OK)
+        return true;
+    else
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:121]: `hr == S_OK`

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 16]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+        std::cout << "Failed to call/handle the async complete handler, gracefully cancel the channelOperation" << std::endl;
+        channelOperation.Cancel();
+        hr = ERROR_TIMEOUT;
     }
-    catch (winrt::hresult_error const& ex)
+    else
     {
-        std::wcout << ex.message().c_str() << std::endl;
+        std::cout << "Channel complete handler has been completed, safely closing the async operation" << std::endl;
+        channelOperation.Close(); // Do not call getresults after this
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:115]: Nit: `GetResults()`

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 17]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    {
+        std::cout << "Failed to call/handle the async complete handler, gracefully cancel the channelOperation" << std::endl;
+        channelOperation2.Cancel();
+        hr = ERROR_TIMEOUT;
+    }
+
+    if (hr == WPN_E_OUTSTANDING_CHANNEL_REQUEST)
+        return true;
+    else
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:211]: Please condense.

File: test/PushNotificationsTests/APITests.cpp [hunk 18]
```diff
@@ -91,13 +91,65 @@ namespace Test::PushNotifications
...
+        {
+            wil::unique_event event = CreateTestEvent(c_testProtocolScheme_Packaged);
+
+            // This is associated protocol for the MSIX installed app for launch.
+            // Use the ://path to define the component you want to test.
+            Uri launchUri{ c_testProtocolScheme_Packaged + L"://ChannelRequestUsingRemoteId" };
+            auto launchResult = Launcher::LaunchUriAsync(launchUri).get();
+            VERIFY_IS_TRUE(launchResult);
+
+            WaitForEvent(event, m_failed);
```
Comment (danielayala94) [inline test/PushNotificationsTests/APITests.cpp:117]: Could all of this be parameterized in one function? All test functions follow the same pattern. The only thing that differs is the test URI string.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 19]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+}
+
+bool MultipleChannelRequestUsingSameRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto hr = S_OK;
+
+    auto channelOperation1 = PushNotificationManager::CreateChannelAsync(winrt::guid(0xa2e4a323, 0xb518, 0x4799, { 0x9e, 0x80, 0x0b, 0x37, 0xae, 0xb0, 0xd2, 0x25 }));
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:129]: I see `channelOperation2` doing something, but I just see this one declared.

If this is your test approach, can you elaborate on the intention here?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 20]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    {
+        std::cout << "Failed to call/handle the async complete handler, gracefully cancel the channelOperation" << std::endl;
+        channelOperation2.Cancel();
+        hr = ERROR_TIMEOUT;
+    }
+
+    if (hr == WPN_E_OUTSTANDING_CHANNEL_REQUEST)
+        return true;
+    else
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:167]: Please condense.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 21]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    {
+        std::cout << "Failed to call/handle the async complete handler, gracefully cancel the channelOperation" << std::endl;
+        channelOperation2.Cancel();
+        channelOperationResult2 = ERROR_TIMEOUT;
+    }
+
+    if ((channelOperationResult2 == WPN_E_OUTSTANDING_CHANNEL_REQUEST) && (channelOperationResult3 == WPN_E_OUTSTANDING_CHANNEL_REQUEST))
+        return true;
+    else
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:294]: Please condense.

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 22]
```diff
@@ -1,21 +1,22 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in the project root for license information.
-#include "pch.h"
+﻿#include "pch.h"
 #include <testdef.h>
-
-#include <wil/win32_helpers.h>
 #include <iostream>
+#include <sstream>
+#include <regex>
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:5]: What APIs are justifying this inclusion?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 23]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+    else
+        return false;
+}
+
+bool MultipleChannelRequestUsingMultipleRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto hr = S_OK;
+
+    auto channelOperation1 = PushNotificationManager::CreateChannelAsync(winrt::guid(0xa2e4a323, 0xb518, 0x4799, { 0x9e, 0x80, 0x0b, 0x37, 0xae, 0xb0, 0xd2, 0x25 }));
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:175]: We are the declaring the guid over and over again. Don't we want to define one, and just reference it every time we want to request a channel?

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 24]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
+        return MultipleChannelRequestUsingSameRemoteId();
+
+    case UnitTest::multipleChannelRequestUsingMultipleRemoteId:
+        return MultipleChannelRequestUsingMultipleRemoteId();
+
+    case UnitTest::threeChannelRequestUsingSameRemoteId:
+        return ThreeChannelRequestUsingSameRemoteId();
+
+    default:
+        return false;
```
Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:317]: Nit: Add one indentation.

Comment (danielayala94) [inline test/TestApps/PushNotificationsTestApp/main.cpp:317]: What if we print the unrecognized test name?

**Review by danielayala94 — CHANGES_REQUESTED** [review]

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 11 above)
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp:85]: if you see the product code, S_OK is returned when we have a valid channel. So just checking for HRESULT should be fine here.

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 12 above)
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp:86]: Actually std::cout are unneccessary, got rid of them. Unit test has to tell us whether it is a success/failure.

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 4 above)
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp]: Done

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 5 above)
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp]: Introduced a global remoteId, we don't need these comments now

File: test/TestApps/PushNotificationsTestApp/main.cpp [hunk 25]
```diff
@@ -24,49 +25,328 @@ void SignalPhase(const std::wstring& phaseEventName)
...
     }
 }
 
 int main()
 {
-    SignalPhase(c_testProtocolScheme_Packaged);
+    initUnitTestMapping();
+    bool succeeded = true;
+    auto args = winrt::Microsoft::ApplicationModel::Activation::AppLifecycle::GetActivatedEventArgs(); // mock name for now
```
[the remark this answers is not in the document]
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp:282]: no all the test code is in this branch.

File: test/TestApps/PushNotificationsTestApp/main.cpp (hunk 7 above)
Comment (sharath2727) [reply test/TestApps/PushNotificationsTestApp/main.cpp]: All these error messages aren't needed. So getting rid of it. Also, to answer your question, InProgressRetry means we are retrying channel request again and there will be a delay to fetch it.

File: test/PushNotificationsTests/APITests.cpp [hunk 26]
```diff
@@ -4,32 +4,21 @@
 #include "pch.h"
 #include <testdef.h>
 #include "Shared.h"
-#include "winrt/Windows.Foundation.h"
 
 using namespace WEX::Common;
```
Comment (hulumane) [inline test/PushNotificationsTests/APITests.cpp]: Can you also please fix the bug in IsPackagedProcess() in this review. It incorrectly assumes that the error code is an HRESULT

File: test/PushNotificationsTests/APITests.cpp [hunk 27]
```diff
@@ -121,8 +110,8 @@ namespace Test::PushNotifications
 
             // This is associated protocol for the MSIX installed app for launch.
             // Use the ://path to define the component you want to test.
-            winrt::Windows::Uri launchUri{ c_testProtocolScheme_Packaged + L"://ChannelRequestUsingRemoteId" };
-            auto launchResult = winrt::Windows::Launcher::LaunchUriAsync(launchUri).get();
+            Uri launchUri{ c_testProtocolScheme_Packaged + L"://ChannelRequestUsingRemoteId" };
+            auto launchResult = Launcher::LaunchUriAsync(launchUri).get();
             VERIFY_IS_TRUE(launchResult);
 
             WaitForEvent(event, m_failed);
```
Comment (hulumane) [inline test/PushNotificationsTests/APITests.cpp:117]: Should m_failed be verified?

**Comment by sharath2727** [comment]
This content is now tracked as part of: https://github.com/microsoft/ProjectReunion/pull/821/files

**Pull request closed without merging**